File: MySpace/src/app/services/job.service.ts
import {JobModel} from "../models/job.model";
import {HttpClient, HttpHeaders} from "@angular/common/http";
import {environment} from "../../environments/environment";
import {Observable} from "rxjs";
import {Injectable} from "@angular/core";
import {FormGroup} from "@angular/forms";
import {AuthService} from "./auth.service";

@Injectable({
  providedIn: 'root',
})
export class JobService {
  constructor(private http: HttpClient,
              private authService: AuthService) {
  }

  private static readonly ENDPOINT = environment.apiUrl + "/job"

  getJobs() : Observable<JobModel[]> {
    return this.http.get<JobModel[]>(JobService.ENDPOINT);
  }

  getJob(id: string) : Observable<JobModel> {
    return this.http.get<JobModel>(JobService.ENDPOINT + "/" + id);
  }

  getCurrentJob() : Observable<JobModel> {
    return this.http.get<JobModel>(JobService.ENDPOINT + "/current");
  }

  addJob(form: FormGroup) : Observable<JobModel> {
    return this.http.post<JobModel>(JobService.ENDPOINT, JSON.stringify(form.value), this.getHttpOptions());
  }

  updateJob(id: string, form: FormGroup) : Observable<JobModel> {
    return this.http.put<JobModel>(JobService.ENDPOINT + "/" + id, JSON.stringify(form.value), this.getHttpOptions());
  }

  deleteJob(id: string) : Observable<any> {
    return this.http.delete(JobService.ENDPOINT + "/" + id, this.getHttpOptions());
  }

  private getHttpOptions() {
    let token = this.authService.getToken();
    if (token == null) {
      throw new Error("Unauthorized.")
    }

    return {
      headers: new HttpHeaders({
        "Content-Type": "application/json",
        "Authorization": "Bearer " + token
      })
    }
  }
}

File: MySpace.Api/MySpace.Api.Application/Services/JobService.cs
using MySpace.Api.Application.Exceptions;
using MySpace.Api.Application.Persistence;
using MySpace.Api.Domain.Models;

namespace MySpace.Api.Application.Services;

public class JobService : IJobService
{
    private readonly JobRepository _jobRepository;

    public JobService(JobRepository jobRepository)
    {
        _jobRepository = jobRepository;
    }

    public List<Job> GetJobs()
    {
        return _jobRepository.GetJobs();
    }

    public Job? GetCurrentJob()
    {
        return _jobRepository.GetCurrentJob();
    }

    public Job GetJob(int id)
    {
        ThrowExceptionIfJobNotFound(id);
        return _jobRepository.GetJob(id);
    }

    public Job AddJob(Job job)
    {
        return _jobRepository.AddJob(job);
    }

    public Job EditJob(int id, Job job)
    {
        ThrowExceptionIfJobNotFound(id);
        return _jobRepository.EditJob(id, job);
    }

    public void DeleteJob(int id)
    {
        ThrowExceptionIfJobNotFound(id);
        _jobRepository.DeleteJob(id);
    }
    
    private void ThrowExceptionIfJobNotFound(int id)
    {
        if (!_jobRepository.JobExists(id))
        {
            throw new JobNotFoundException(id.ToString());
        }
    }
}